File: backend/services/personality_engine.py
from __future__ import annotations

import logging
import time
from typing import Dict, List, Optional

from backend.config import PERSONALITY_TRAITS, TRAIT_BADGE_THRESHOLD
from backend.models.speaker_profile import SpeakerProfile

logger = logging.getLogger(__name__)

SIGNAL_WEIGHTS = {
    "emotion": {
        "sarcastic": {"sarcasm": 0.8},
        "dramatic": {"excitement": 0.5, "anger": 0.6, "surprise": 0.3},
        "curious": {"curiosity": 0.7},
        "supportive": {"excitement": 0.2},
        "skeptical": {"anger": 0.2},
    },
    "intent": {
        "curious": {"question": 0.6},
        "secretive": {"gossip": 0.7},
        "supportive": {"greeting": 0.2},
    },
    "message_length": {
        "dramatic": {"long": 0.3},
        "curious": {"short_question": 0.2},
    },
}

SUPPORTIVE_PATTERNS = [
    "hope",
    "care",
    "love you",
    "proud",
    "great job",
    "you got this",
    "don't worry",
    "here for you",
    "support",
    "believe in",
]

SKEPTICAL_PATTERNS = [
    "doubt",
    "don't think",
    "unlikely",
    "not sure about",
    "really?",
    "hard to believe",
    "suspicious",
    "don't trust",
    "seems fake",
    "don't buy",
]


class PersonalityEngine:
    def __init__(self) -> None:
        self._profiles: Dict[str, SpeakerProfile] = {}

    def update_trait_score(
        self,
        profile: SpeakerProfile,
        trait: str,
        signal_weight: float,
    ) -> float:
        current = profile.trait_scores.get(trait, 0.0)
        updated = current * 0.9 + signal_weight * 0.1
        profile.trait_scores[trait] = updated
        return updated

    def analyze_signals(
        self,
        profile: SpeakerProfile,
        emotion: str,
        sarcasm: bool,
        intent: str,
        text: str,
        topic_category: str,
    ) -> Dict[str, float]:
        text_lower = text.lower()
        msg_len = len(text)

        # Sarcasm signal
        if sarcasm:
            self.update_trait_score(profile, "sarcastic", 1.0)
        else:
            self.update_trait_score(profile, "sarcastic", 0.0)

        # Emotion-based signals
        emotion_weights = SIGNAL_WEIGHTS["emotion"]
        for trait, emotion_map in emotion_weights.items():
            weight = emotion_map.get(emotion, 0.0)
            if weight > 0:
                self.update_trait_score(profile, trait, weight)

        # Intent-based signals
        intent_weights = SIGNAL_WEIGHTS["intent"]
        for trait, intent_map in intent_weights.items():
            weight = intent_map.get(intent, 0.0)
            if weight > 0:
                self.update_trait_score(profile, trait, weight)

        # Curious from questions
        if intent == "question":
            self.update_trait_score(profile, "curious", 0.8)
        else:
            self.update_trait_score(profile, "curious", 0.0)

        # Dramatic from long emotional messages
        if msg_len > 100 and emotion in ("excitement", "anger", "surprise"):
            self.update_trait_score(profile, "dramatic", 0.9)
        elif emotion in ("excitement", "anger"):
            self.update_trait_score(profile, "dramatic", 0.5)
        else:
            self.update_trait_score(profile, "dramatic", 0.0)

        # Supportive from keyword matching
        supportive_hits = sum(
            1 for p in SUPPORTIVE_PATTERNS if p in text_lower
        )
        if supportive_hits > 0:
            weight = min(supportive_hits * 0.3, 1.0)
            self.update_trait_score(profile, "supportive", weight)
        else:
            self.update_trait_score(profile, "supportive", 0.0)

        # Secretive from gossip intent/topic
        if intent == "gossip" or topic_category == "gossip_tea":
            self.update_trait_score(profile, "secretive", 0.8)
        else:
            self.update_trait_score(profile, "secretive", 0.0)

        # Skeptical from patterns
        skeptical_hits = sum(
            1 for p in SKEPTICAL_PATTERNS if p in text_lower
        )
        if skeptical_hits > 0:
            weight = min(skeptical_hits * 0.3, 1.0)
            self.update_trait_score(profile, "skeptical", weight)
        else:
            self.update_trait_score(profile, "skeptical", 0.0)

        # Update emotion distribution
        total = profile.conversation_count or 1
        for emo_key in profile.emotion_distribution:
            current = profile.emotion_distribution[emo_key]
            detected = 1.0 if emo_key == emotion else 0.0
            profile.emotion_distribution[emo_key] = (
                current + (detected - current) / total
            )

        # Update topics
        profile.top_topics[topic_category] = (
            profile.top_topics.get(topic_category, 0) + 1
        )

        # Increment conversation count
        profile.conversation_count += 1
        profile.updated_at = time.time()

        return dict(profile.trait_scores)

    def get_or_create_profile(
        self, speaker_id: str, name: Optional[str] = None
    ) -> SpeakerProfile:
        if speaker_id not in self._profiles:
            self._profiles[speaker_id] = SpeakerProfile(
                id=speaker_id, name=name
            )
        return self._profiles[speaker_id]

    def get_profile(self, speaker_id: str) -> Optional[SpeakerProfile]:
        return self._profiles.get(speaker_id)

    def get_all_profiles(self) -> List[SpeakerProfile]:
        return list(self._profiles.values())

    def get_badge_traits(self, speaker_id: str) -> List[str]:
        profile = self._profiles.get(speaker_id)
        if profile is None:
            return []
        return [
            trait
            for trait, score in profile.trait_scores.items()
            if score >= TRAIT_BADGE_THRESHOLD
        ]

    def recompute_from_history(
        self, speaker_id: str, messages: List[Dict]
    ) -> Optional[SpeakerProfile]:
        profile = self._profiles.get(speaker_id)
        if profile is None:
            return None

        profile.trait_scores = {t: 0.0 for t in PERSONALITY_TRAITS}
        profile.emotion_distribution = {
            "sarcasm": 0.0,
            "excitement": 0.0,
            "anger": 0.0,
            "curiosity": 0.0,
            "amusement": 0.0,
            "surprise": 0.0,
        }
        profile.top_topics = {}
        profile.conversation_count = 0

        for msg in messages:
            self.analyze_signals(
                profile=profile,
                emotion=msg.get("emotion", "excitement"),
                sarcasm=msg.get("emotion") == "sarcasm",
                intent=msg.get("intent", "statement"),
                text=msg.get("text", ""),
                topic_category=msg.get("topic", "general"),
            )

        profile.updated_at = time.time()
        self._profiles[speaker_id] = profile
        return profile


personality_engine = PersonalityEngine()

File: GossipAppFixed/src/modules/personality/PersonalityEngine.ts
import {
  FriendProfile,
  ConversationAnalysis,
  ReplySuggestion,
  PersonalityTrait,
  Emotion,
} from './types';
import { detectEmotion } from './EmotionDetector';
import { extractTopic } from './TopicExtractor';
import { generateSuggestions } from './ResponseGenerator';

function detectIntent(text: string): string {
  const lower = text.toLowerCase();
  if (/\b(hi|hey|hello|sup|what'?s up|yo)\b/i.test(lower)) return 'greeting';
  if (/\?/.test(text)) return 'question';
  if (/heard|apparently|did you know|rumor|secret|tea|drama/i.test(lower)) return 'gossip';
  return 'statement';
}

function detectTone(emotion: Emotion, sarcasm: boolean): string {
  if (sarcasm) return 'sarcastic';
  switch (emotion) {
    case 'excitement': return 'enthusiastic';
    case 'anger': return 'frustrated';
    case 'curiosity': return 'inquisitive';
    case 'amusement': return 'playful';
    case 'surprise': return 'shocked';
    default: return 'casual';
  }
}

function getSpeechStyle(messages: string[]): string[] {
  if (messages.length === 0) return [];

  const styles: string[] = [];
  const avgLength = messages.reduce((sum, m) => sum + m.length, 0) / messages.length;

  if (avgLength < 20) styles.push('short_messages');
  else if (avgLength > 80) styles.push('long_messages');

  const questionRatio = messages.filter(m => m.includes('?')).length / messages.length;
  if (questionRatio > 0.3) styles.push('rhetorical_questions');

  const emojiPattern = /[\u{1F600}-\u{1F64F}\u{1F300}-\u{1F5FF}\u{1F680}-\u{1F6FF}\u{1F1E0}-\u{1F1FF}\u{2600}-\u{26FF}\u{2700}-\u{27BF}]/u;
  const emojiRatio = messages.filter(m => emojiPattern.test(m)).length / messages.length;
  if (emojiRatio > 0.3) styles.push('emoji_heavy');

  const capsRatio = messages.filter(m => {
    const words = m.replace(/[^a-zA-Z\s]/g, '').trim().split(/\s+/);
    return words.filter(w => w.length > 1 && w === w.toUpperCase()).length >= 2;
  }).length / messages.length;
  if (capsRatio > 0.2) styles.push('caps_shouter');

  const exclamRatio = messages.filter(m => /!{2,}/.test(m)).length / messages.length;
  if (exclamRatio > 0.3) styles.push('exclamation_heavy');

  return styles;
}

class PersonalityEngine {
  private profiles: Map<string, FriendProfile> = new Map();
  private analysisHistory: ConversationAnalysis[] = [];
  private messageHistory: Map<string, string[]> = new Map();

  analyzeMessage(senderId: string, senderName: string, text: string): ConversationAnalysis {
    const emotionResult = detectEmotion(text);
    const topicResult = extractTopic(text);
    const intent = detectIntent(text);
    const tone = detectTone(emotionResult.emotion, emotionResult.sarcasm);

    const analysis: ConversationAnalysis = {
      speakerId: senderId,
      text,
      intent,
      topic: topicResult.topic,
      emotion: emotionResult.emotion,
      sarcasm: emotionResult.sarcasm,
      tone,
      confidence: emotionResult.confidence,
      timestamp: Date.now(),
    };

    this.analysisHistory.push(analysis);
    this.updateProfile(senderId, senderName, analysis, text);

    return analysis;
  }

  private updateProfile(
    senderId: string,
    senderName: string,
    analysis: ConversationAnalysis,
    rawText: string,
  ): void {
    let profile = this.profiles.get(senderId);

    if (!profile) {
      profile = {
        id: senderId,
        name: senderName,
        conversationCount: 0,
        topics: new Map(),
        personalityTraits: new Map(),
        speechStyle: [],
        emotionDistribution: new Map(),
        lastUpdated: Date.now(),
      };
    }

    profile.conversationCount++;
    profile.lastUpdated = Date.now();

    // Update topics
    const currentCount = profile.topics.get(analysis.topic) || 0;
    profile.topics.set(analysis.topic, currentCount + 1);

    // Update emotion distribution
    const totalMessages = profile.conversationCount;
    for (const emotion of ['excitement', 'sarcasm', 'anger', 'curiosity', 'surprise', 'amusement', 'neutral'] as Emotion[]) {
      const current = profile.emotionDistribution.get(emotion) || 0;
      const detected = analysis.emotion === emotion ? 1 : 0;
      profile.emotionDistribution.set(emotion, current + (detected - current) / totalMessages);
    }

    // Update personality traits with exponential moving average
    this.updateTrait(profile, 'sarcastic', analysis.sarcasm ? 1 : 0);
    this.updateTrait(profile, 'curious', analysis.intent === 'question' ? 1 : 0);
    this.updateTrait(profile, 'dramatic', analysis.emotion === 'excitement' || analysis.emotion === 'anger' ? 1 : 0);
    this.updateTrait(profile, 'supportive', /hope|care|love you|proud|great job/i.test(rawText) ? 1 : 0);
    this.updateTrait(profile, 'secretive', analysis.intent === 'gossip' ? 1 : 0);
    this.updateTrait(profile, 'skeptical', /doubt|don't think|unlikely|not sure about/i.test(rawText) ? 1 : 0);

    // Track messages for speech style analysis
    const messages = this.messageHistory.get(senderId) || [];
    messages.push(rawText);
    this.messageHistory.set(senderId, messages);
    profile.speechStyle = getSpeechStyle(messages);

    this.profiles.set(senderId, profile);
  }

  private updateTrait(profile: FriendProfile, trait: PersonalityTrait, detected: number): void {
    const current = profile.personalityTraits.get(trait) || 0;
    const updated = current * 0.9 + detected * 0.1;
    profile.personalityTraits.set(trait, updated);
  }

  getProfile(speakerId: string): FriendProfile | undefined {
    return this.profiles.get(speakerId);
  }

  getAllProfiles(): FriendProfile[] {
    return Array.from(this.profiles.values());
  }

  getSuggestions(analysis: ConversationAnalysis): ReplySuggestion[] {
    const profile = this.profiles.get(analysis.speakerId);
    return generateSuggestions(analysis, profile);
  }
}

export const personalityEngine = new PersonalityEngine();
export default PersonalityEngine;
